lib/processors/pr0nprocessor.py

```diff
@@ -0,0 +1,264 @@
+from nearpy.hashes import RandomBinaryProjections
+import os
+import pickle
+import itertools as IT
+from collections import Counter
+from operator import itemgetter
+
+from ..config import CONFIG
+from .lib.utils import process_api_handler
+from .lib.baseprocessor import BaseProcessor
+
+
+class Pr0nProcessor(BaseProcessor):
+    name = 'pr0n_processor'
+    data = {}
+
+    def __init__(self):
+        super().__init__()
+        try:
+            fname = './data/pr0n/backend.pkl'
+            with open(fname, 'rb') as fd:
+                self.pr0n_engine = pickle.load(fd)
+            self.logger.info("Loaded backend from save state")
+        except (IOError, ValueError):
+            self.logger.info("Recreating backend engine")
+            self.pr0n_engine = self.create_engine()
+            self.read_pr0n()
+            with open(fname, 'wb+') as fd:
+                pickle.dump(self.pr0n_engine, fd)
+
+    def _get_img(self, img_hash):
+        fname = "data/pr0n/backend/{0[0]}/{0[1]}/{0}.pkl".format(img_hash)
+        with open(fname, 'rb') as fd:
+            return pickle.load(fd)
+
+    def create_engine(self):
+        rbp = RandomBinaryProjections('rbp', 8)
+        return Engine(128, lshashes=[rbp])
+
+    def read_pr0n(self):
+        num_added = 0
+        for dirname, _, filenames in os.walk('./data/pr0n/backend/'):
+            if filenames:
+                for filename in filenames:
+                    if not filename.endswith('pkl'):
+                        continue
+                    filepath = os.path.join(dirname, filename)
+                    filehash = filename.split('.')[0]
+                    with open(filepath, 'rb') as fd:
+                        data = pickle.load(fd)
+                        if data['pose'] != 0:
+                            continue
+                        self.pr0n_engine.store_vector(data['face_hash'],
+                                                      filehash)
+                        num_added += 1
+        self.logger.info("Added backend pictures: {}".format(num_added))
+        if not num_added:
+            self.logger.error("PRON: No pictures added... try running "
+                              "./scripts/create_pr0n_database.py")
+
+    def filter_photos(self, photos):
+        for photo in photos:
+            if not photo.get('faces'):
+                continue
+            elif not any(f.get('pose') == 0 for f in photo['faces']):
+                continue
+            elif not any(f.get('face_hash', None) is not None
+                         for f in photo['faces']):
+                continue
+            yield photo
+
+    @gen.coroutine
+    def process(self, user_data):
+        """
+        Go through all the tagged facebook pictures from the user and creates a
+        nearpy structure for easy similarity lookup
+        """
+        self.logger.info("Processing user: {}".format(user_data.userid))
+        if not user_data.data.get('fbphotos'):
```

Comment:
Is this a facebook only game since it is the only service with tagging or is gphotos or insta possible too?